engine.go

```diff
@@ -0,0 +1,247 @@
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var runOutput, runErr bytes.Buffer
+	cmdUser.Start()
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go piper.PipeOutput(&wg, outUser, os.Stdout, &runOutput)
+	go piper.PipeOutput(&wg, errUser, os.Stdout, &runErr)
+
+	exitErr := cmdUser.Wait()
+	wg.Wait()
+	prepLog, err := ioutil.ReadFile(tempDir + "/prepare.log")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var stats syscall.Rusage
+	statsData, err := ioutil.ReadFile(tempDir + "/stats.log")
+	if err != nil {
+		log.Print(err)
+	} else {
+		err = json.Unmarshal(statsData, &stats)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	piper.LogRunComplete(key, build, "", runOutput.String(), "", exitErr, string(prepLog), stats)
+
+	var toSend = make(map[string]string)
+	toSend["run"] = runOutput.String()
+	toSend["err"] = runErr.String()
+
+	json, err := json.Marshal(toSend)
+	if err != nil {
+		http.Error(w, "Json marshal err: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.Write(json)
+}
+
+func createExecFile(tmpDir, lang, codeBase string) error {
+	f, err := os.Create(tmpDir + "/" + fileNames[lang])
+	if err != nil {
+		return err
+	}
+	f.WriteString(codeBase)
+	f.Close()
+	return nil
+}
+
+func createConfigurationFile(tempDir, lang string) error {
+	if err := copyFileContents(languages[lang], tempDir, configFileName); err != nil {
+		return err
+	}
+	return nil
+}
+
+func createTempDir() (string, string, error) {
+	id, err := uuid.V4()
+	if err != nil {
+		return "", "", err
+	}
+	tempDir := homeDir + id.String()
+	os.MkdirAll(tempDir, 0777)
+
+	return tempDir, dockerHomeDir + id.String(), nil
+}
+
+// copyFileContents copies the contents of the file named src to the file named
+// by dst. The file will be created if it does not already exist. If the
+// destination file exists, all it's contents will be replaced by the contents
+// of the source file.
+func copyFileContents(src, dst, fileName string) (err error) {
+	in, err := os.Open(src)
+	if err != nil {
+		return
+	}
+	dst = dst + "/" + fileName
+	defer in.Close()
+	out, err := os.Create(dst)
+	if err != nil {
+		return
+	}
+	defer func() {
+		cerr := out.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+	if _, err = io.Copy(out, in); err != nil {
+		return
+	}
+	err = out.Sync()
+	return
+}
+
+func main() {
+	if runtime.GOOS == "windows" {
+		u, err := user.Current()
+		if err != nil {
+			fmt.Println(err.Error())
+			return
+		}
+		homeDir = u.HomeDir
+		homeDir = strings.Replace(homeDir, "\\", "/", -1) + "/tmp/"
+		if !strings.HasPrefix(u.HomeDir, "C:\\Users\\") {
+			fmt.Println("Your home path should begin with C:\\Users\\ because you are on windows and that`s the only boot2docker shared folder")
+			return
+		}
+		dockerHomeDir = "/" + homeDir
+		dockerHomeDir = strings.Replace(dockerHomeDir, ":", "", 1)
+		dockerHomeDir = strings.Replace(dockerHomeDir, "C", "c", 1)
+		dockerHomeDir = strings.Replace(dockerHomeDir, "\\", "/", -1)
+	} else {
+		homeDir = os.Getenv("HOME") + "/tmp/"
```

Comment:
This is a no-go. We won't just fuck around with our `$HOME`s. Use [`ioutil.TempDir`](https://golang.org/pkg/io/ioutil/#TempDir) in in that case.